corgi/api/views.py

```diff
@@ -380,12 +381,49 @@ class ComponentViewSet(ReadOnlyModelViewSet):  # TODO: TagViewMixin disabled unt
     filterset_class = ComponentFilter
     lookup_url_kwarg = "uuid"
 
+    def get_queryset(self):
+        mode = self.request.query_params.get("mode")
+        if mode == "latest":
+            # 'latest' mode only relevent in terms of a specific offering/product
+            # TODO - subject of a near-future PR will handle instances of when we have
+            #      multiple root components with the same component name yet different versions.
+            ofuri = self.request.query_params.get("ofuri")
+            if ofuri:
+                component_types = Q(type=Component.Type.SRPM) | Q(
+                    type=Component.Type.CONTAINER_IMAGE
+                )
+                results = (
+                    Component.objects.filter(
+                        component_types,
+                        product_streams__overlap=[ofuri],
+                        software_build__completion_time__isnull=False,
+                    )
+                    .annotate(
+                        latest=Subquery(
+                            Component.objects.filter(
+                                component_types,
+                                name=OuterRef("name"),
+                                product_streams__overlap=[ofuri],
+                                software_build__completion_time__isnull=False,
+                            )
+                            .order_by("software_build__completion_time")
+                            .values("uuid")[:1]
+                        )
+                    )
+                    .filter(
+                        uuid=F("latest"),
```

Comment:
I haven't tested this query, but does this actually work for lists of components? If e.g. ofuri:rhel-8 ships emacs-1, emacs-2, emacs-3, openssl-4, openssl-5, will this spit out [emacs-3, openssl-5]? The way I'm reading this query suggest it would only spit out "openssl-5".

It would be nice to have a test to go along with this.

Reply:
yes it works as you describe ([emacs-3, openssl-5]) - can do a demo tmrw to show you behaviour and yes will add a test ;)

Reply:
Also, it just occurred to me that a more readable query might start by finding the distinct builds that ship latest components and then filter Components on that, something like:

```
Component.objects.filter(
    software_build__in=SoftwareBuild.objects.filter(completion_time__isnull=False)
   .order_by("components__name", "-completion_time")
   .distinct("components__name")
).distinct()
```

(I left out the other ofuri and type filters.)